File: P/1706_Where_Will_the_Ball_Fall.java
class Solution {
    int[][] grid;
    public int[] findBall(int[][] grid) {
        this.grid = grid;
        int[] ans = new int[grid[0].length];
        
        for(int i = 0; i < grid[0].length; i++)
            ans[i] = roll(i);

        return ans;
    }
    
    private int roll(int n) {
        int cur = n;
        int prev;
        for(int j = 0; j < this.grid.length; j++) {
            prev = this.grid[j][cur];
            cur += prev;
            
            if(cur < 0 || cur >= grid[0].length)
                return -1;
            
            if(prev == 1 && this.grid[j][cur] == -1)
                return -1;
            if(prev == -1 && this.grid[j][cur] == 1)
                return -1;
        }
        return cur;
    }
}

File: 주영/Graph/1706. Where Will the Ball Fall.js
const solution = grid => {
  let answer = [];
  for (let i = 0; i < grid[0].length; i++) answer[i] = roll(i, grid);

  return answer;
};

const roll = (i, grid) => {
  let prev;

  for (let j = 0; j < grid.length; j++) {
    prev = grid[j][i];
    i += prev;

    if (i < 0 || i >= grid[0].length) return -1;
    if (prev === 1 && grid[j][i] === -1) return -1;
    if (prev === -1 && grid[j][i] === 1) return -1;
  }
  return i;
};

let grid = [
  [1, 1, 1, -1, -1],
  [1, 1, 1, -1, -1],
  [-1, -1, -1, 1, 1],
  [1, 1, 1, 1, -1],
  [-1, -1, -1, -1, -1],
];

solution(grid);
